// @midscene/core/dist/lib/env.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});



































var _chunkI2I7OU2Ejs = require('./chunk-I2I7OU2E.js');




































exports.ANTHROPIC_API_KEY = _chunkI2I7OU2Ejs.ANTHROPIC_API_KEY; exports.AZURE_OPENAI_API_VERSION = _chunkI2I7OU2Ejs.AZURE_OPENAI_API_VERSION; exports.AZURE_OPENAI_DEPLOYMENT = _chunkI2I7OU2Ejs.AZURE_OPENAI_DEPLOYMENT; exports.AZURE_OPENAI_ENDPOINT = _chunkI2I7OU2Ejs.AZURE_OPENAI_ENDPOINT; exports.AZURE_OPENAI_KEY = _chunkI2I7OU2Ejs.AZURE_OPENAI_KEY; exports.MATCH_BY_POSITION = _chunkI2I7OU2Ejs.MATCH_BY_POSITION; exports.MIDSCENE_API_TYPE = _chunkI2I7OU2Ejs.MIDSCENE_API_TYPE; exports.MIDSCENE_AZURE_OPENAI_INIT_CONFIG_JSON = _chunkI2I7OU2Ejs.MIDSCENE_AZURE_OPENAI_INIT_CONFIG_JSON; exports.MIDSCENE_AZURE_OPENAI_SCOPE = _chunkI2I7OU2Ejs.MIDSCENE_AZURE_OPENAI_SCOPE; exports.MIDSCENE_CACHE = _chunkI2I7OU2Ejs.MIDSCENE_CACHE; exports.MIDSCENE_DANGEROUSLY_PRINT_ALL_CONFIG = _chunkI2I7OU2Ejs.MIDSCENE_DANGEROUSLY_PRINT_ALL_CONFIG; exports.MIDSCENE_DEBUG_AI_PROFILE = _chunkI2I7OU2Ejs.MIDSCENE_DEBUG_AI_PROFILE; exports.MIDSCENE_DEBUG_AI_RESPONSE = _chunkI2I7OU2Ejs.MIDSCENE_DEBUG_AI_RESPONSE; exports.MIDSCENE_DEBUG_MODE = _chunkI2I7OU2Ejs.MIDSCENE_DEBUG_MODE; exports.MIDSCENE_LANGSMITH_DEBUG = _chunkI2I7OU2Ejs.MIDSCENE_LANGSMITH_DEBUG; exports.MIDSCENE_MODEL_NAME = _chunkI2I7OU2Ejs.MIDSCENE_MODEL_NAME; exports.MIDSCENE_OPENAI_INIT_CONFIG_JSON = _chunkI2I7OU2Ejs.MIDSCENE_OPENAI_INIT_CONFIG_JSON; exports.MIDSCENE_OPENAI_SOCKS_PROXY = _chunkI2I7OU2Ejs.MIDSCENE_OPENAI_SOCKS_PROXY; exports.MIDSCENE_REPORT_TAG_NAME = _chunkI2I7OU2Ejs.MIDSCENE_REPORT_TAG_NAME; exports.MIDSCENE_USE_ANTHROPIC_SDK = _chunkI2I7OU2Ejs.MIDSCENE_USE_ANTHROPIC_SDK; exports.MIDSCENE_USE_AZURE_OPENAI = _chunkI2I7OU2Ejs.MIDSCENE_USE_AZURE_OPENAI; exports.MIDSCENE_USE_DOUBAO_VISION = _chunkI2I7OU2Ejs.MIDSCENE_USE_DOUBAO_VISION; exports.MIDSCENE_USE_QWEN_VL = _chunkI2I7OU2Ejs.MIDSCENE_USE_QWEN_VL; exports.MIDSCENE_USE_VLM_UI_TARS = _chunkI2I7OU2Ejs.MIDSCENE_USE_VLM_UI_TARS; exports.MIDSCENE_USE_VL_MODEL = _chunkI2I7OU2Ejs.MIDSCENE_USE_VL_MODEL; exports.OPENAI_API_KEY = _chunkI2I7OU2Ejs.OPENAI_API_KEY; exports.OPENAI_BASE_URL = _chunkI2I7OU2Ejs.OPENAI_BASE_URL; exports.OPENAI_MAX_TOKENS = _chunkI2I7OU2Ejs.OPENAI_MAX_TOKENS; exports.OPENAI_USE_AZURE = _chunkI2I7OU2Ejs.OPENAI_USE_AZURE; exports.allAIConfig = _chunkI2I7OU2Ejs.allAIConfig; exports.getAIConfig = _chunkI2I7OU2Ejs.getAIConfig; exports.getAIConfigInBoolean = _chunkI2I7OU2Ejs.getAIConfigInBoolean; exports.getAIConfigInJson = _chunkI2I7OU2Ejs.getAIConfigInJson; exports.overrideAIConfig = _chunkI2I7OU2Ejs.overrideAIConfig; exports.vlLocateMode = _chunkI2I7OU2Ejs.vlLocateMode;
