// @midscene/core/dist/lib/utils.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});
















var _chunkQ4DRRZOEjs = require('./chunk-Q4DRRZOE.js');
require('./chunk-I2I7OU2E.js');

















exports.getLogDir = _chunkQ4DRRZOEjs.getLogDir; exports.getLogDirByType = _chunkQ4DRRZOEjs.getLogDirByType; exports.getTmpDir = _chunkQ4DRRZOEjs.getTmpDir; exports.getTmpFile = _chunkQ4DRRZOEjs.getTmpFile; exports.getVersion = _chunkQ4DRRZOEjs.getVersion; exports.groupedActionDumpFileExt = _chunkQ4DRRZOEjs.groupedActionDumpFileExt; exports.overlapped = _chunkQ4DRRZOEjs.overlapped; exports.replaceStringWithFirstAppearance = _chunkQ4DRRZOEjs.replaceStringWithFirstAppearance; exports.replacerForPageObject = _chunkQ4DRRZOEjs.replacerForPageObject; exports.reportHTMLContent = _chunkQ4DRRZOEjs.reportHTMLContent; exports.setLogDir = _chunkQ4DRRZOEjs.setLogDir; exports.sleep = _chunkQ4DRRZOEjs.sleep; exports.stringifyDumpData = _chunkQ4DRRZOEjs.stringifyDumpData; exports.uploadTestInfoToServer = _chunkQ4DRRZOEjs.uploadTestInfoToServer; exports.writeDumpReport = _chunkQ4DRRZOEjs.writeDumpReport; exports.writeLogFile = _chunkQ4DRRZOEjs.writeLogFile;
