// @ui-tars/shared/dist/constants/vlm.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var vlm_exports = {};
__export(vlm_exports, {
  IMAGE_PLACEHOLDER: () => IMAGE_PLACEHOLDER,
  MAX_IMAGE_LENGTH: () => MAX_IMAGE_LENGTH,
  MAX_LOOP_COUNT: () => MAX_LOOP_COUNT,
  VlmModeEnum: () => VlmModeEnum,
  VlmModeEnumOptions: () => VlmModeEnumOptions
});
module.exports = __toCommonJS(vlm_exports);
const IMAGE_PLACEHOLDER = "<image>";
const MAX_LOOP_COUNT = 25;
const MAX_IMAGE_LENGTH = 5;
var VlmModeEnum = /* @__PURE__ */ ((VlmModeEnum2) => {
  VlmModeEnum2["Chat"] = "chat";
  VlmModeEnum2["Agent"] = "agent";
  return VlmModeEnum2;
})(VlmModeEnum || {});
const VlmModeEnumOptions = {
  ["agent" /* Agent */]: "Agent \u6A21\u5F0F",
  ["chat" /* Chat */]: "Chat \u6A21\u5F0F"
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  IMAGE_PLACEHOLDER,
  MAX_IMAGE_LENGTH,
  MAX_LOOP_COUNT,
  VlmModeEnum,
  VlmModeEnumOptions
});
